src/kernel/arch/x86/syscalls.zig

```diff
@@ -0,0 +1,173 @@
+const arch = @import("arch.zig");
+const testing = @import("std").testing;
+const assert = @import("std").debug.assert;
+const isr = @import("isr.zig");
+const log = @import("../../log.zig");
+
+pub const INTERRUPT: u16 = 0x80;
+pub const NUM_HANDLERS: u16 = 256;
+
+/// A syscall handler
+pub const SyscallHandler = fn (ctx: *arch.InterruptContext) void;
+/// Errors that syscall utility functions can throw
+pub const SyscallError = error {
+    SyscallExists,
+    InvalidSyscall
+};
+
+/// The array of registered syscalls
+var handlers: [NUM_HANDLERS]?SyscallHandler = []?SyscallHandler{null} ** NUM_HANDLERS;
+
+///
+/// Returns true if the syscall is valid, else false.
+/// A syscall is valid if it's less than NUM_HANDLERS
+///
+/// Arguments:
+///     IN syscall: u32 - The syscall to check
+///
+pub fn isValidSyscall(syscall: u32) bool {
+    return syscall < NUM_HANDLERS;
+}
+
+///
+/// Handle a syscall. Gets the syscall number from eax within the context and calls the registered handler.
+/// If there isn't a registered handler or the syscall is invalid (>= NUM_HANDLERS) then a warning is logged
+///
+/// Arguments:
+///     IN ctx: *arch.InterruptContext - The cpu context when the syscall was triggered. The syscall number is
+///                                     stored in eax.
+///
+fn handle(ctx: *arch.InterruptContext) void {
+    // The syscall number is put in eax
+    const syscall = ctx.eax;
+    if (isValidSyscall(syscall)) {
+        if (handlers[syscall]) |handler| {
+            handler(ctx);
+        } else {
+            log.logWarning("Syscall {} triggered but not registered\n", syscall);
+        }
+    } else {
+        log.logWarning("Syscall {} is invalid\n", syscall);
+    }
+}
+
+///
+/// Register a syscall so it can be called by triggering interrupt 128
+///
+/// Arguments:
+///     IN syscall: u8 - The syscall number to register the handler with
+///     IN handler: SyscallHandler - The handler to register the syscall with
+///
+/// Errors:
+///     SyscallError.InvalidSyscall - If the syscall is invalid (see isValidSyscall)
+///     SyscallError.SyscallExists - If the syscall has already been registered
+///
+pub fn registerSyscall(syscall: u8, handler: SyscallHandler) !void {
+    if (!isValidSyscall(syscall))
+        return SyscallError.InvalidSyscall;
+    if (handlers[syscall]) |_|
+        return SyscallError.SyscallExists;
+    handlers[syscall] = handler;
+}
+
+///
+/// Trigger a syscall. Returns the value put in eax by the syscall handler.
+///
+/// Arguments:
+///     IN syscall: u8 - The syscall to trigger. Is put in eax and must have been registered.
+///     IN args: u32... - The args to send to the syscall. A maximum of 5 are put in ebx, ecx, edx, esi and edi
+///
+pub inline fn doSyscall(comptime syscall: u32, args: ...) u32 {
```

Comment:
I guess this is the point the user land will make a syscall. If so, could this be moved to another file or clearly separate this out so it is clear that this is what the user will call and isn't part of the kernel code.

Reply:
Hmm good question, reckon we should return an error if it's an invalid syscall? How would you suggest we separate it?

Reply:
What would be nice later on is to add this kind of stuff to our own fork of Zig stdlib. That should probably be the final aim.

Reply:
That would be cool. I guess this would be part of the OSs API so a separate file would be good, u can come up with the name ;)